src/Stratis.Bitcoin.IntegrationTests/RPC/ListAddressGroupingsTest.cs

```diff
@@ -0,0 +1,125 @@
+                result.First(a => a.Address == firstChangeAddress.Address).Amount.Should().Be(Money.Coins((decimal)4.9999548 + 5));
+
+                // Send the full balance - 1 from receiver to miner.
+                var balance = this.receiver.FullNode.WalletManager().GetSpendableTransactionsInWallet(walletName).Sum(t => t.Transaction.Amount) - Money.Coins(1);
+                SendCoins(this.receiver, this.miner, balance);
+
+                // Get the change address.
+                receiver_Wallet = this.receiver.FullNode.WalletManager().GetWallet(walletName);
+                var changeAddresses = receiver_Wallet.GetAllAddresses().Where(a => a.IsChangeAddress() && a.Transactions.Any());
+                var secondChangeAddress = receiver_Wallet.GetAllAddresses().First(a => a.IsChangeAddress() && a.Transactions.Any() && a.Address != firstChangeAddress.Address);
+
+                // There should now be 3 items in the listaddressgroupings response and contain another change address.
+                result = await CallListAddressGroupingsAsync();
+                result.Count().Should().Be(3);
+                result.Count(a => a.Address == firstChangeAddress.Address).Should().Be(1);
+                result.Count(a => a.Address == secondChangeAddress.Address).Should().Be(1);
+                result.First(a => a.Address == secondChangeAddress.Address).Amount.Should().Be(Money.Coins((decimal)0.99992520));
+            }
+        }
+
+        private void SendCoins(CoreNode from, CoreNode to, Money coins, HdAddress toAddress = null)
+        {
+            // Get a receive address.
+            if (toAddress == null)
+                toAddress = to.FullNode.WalletManager().GetUnusedAddress(new WalletAccountReference(walletName, accountName));
+
+            // Send 10 coins to node.
+            var transaction = from.FullNode.WalletTransactionHandler().BuildTransaction(WalletTests.CreateContext(from.FullNode.Network, new WalletAccountReference(walletName, accountName), password, toAddress.ScriptPubKey, coins, FeeType.Medium, 10));
+            from.FullNode.NodeService<WalletController>().SendTransaction(new SendTransactionRequest(transaction.ToHex()));
+
+            TestHelper.WaitLoop(() => from.CreateRPCClient().GetRawMempool().Length > 0);
+
+            // Mine the transaction.
+            TestHelper.MineBlocks(this.miner, 10);
+            TestHelper.WaitLoop(() => TestHelper.AreNodesSynced(from, to));
+            TestHelper.WaitLoop(() => to.FullNode.WalletManager().GetSpendableTransactionsInWallet(walletName).Sum(x => x.Transaction.Amount) > 0);
+        }
+
+        private async Task<AddressGroupingModel[]> CallListAddressGroupingsAsync()
+        {
+            RPCClient client = this.receiver.CreateRPCClient();
+            var response = await client.SendCommandAsync(RPCOperations.listaddressgroupings);
+            var result = response.Result.ToObject<AddressGroupingModel[]>();
```

Comment:
I'm running the code manually to verify this; but I don't see how this can give completely correct results. The RPC response is a triply-nested array like this:

```
[
 [
   [
     "mmUToPM2vBb2F1ZNcR86J12GxqAbUJtFCr",
     0.06000000,
     "From Electrum Wallet"
   ]
 ],
 [
   [
     "2MwBERpNv7JJaW4nVUbQncXjFjG1nFKLYEp",
     0.05000000,
     "test"
   ],
   [
     "2N4AMdXitdYhNdjafzHG61uUdrErExSdhKs",
     0.00498305
   ]
 ],
 [
   [
     "2N6xTkaeuKsQSHCortgxiNvsVXj2Gv1Qy85",
     0.05000000,
     "test2"
   ]
 ]
]
```

In other words, it is a list-of-groups. Changing the response to a flat array, as this method seems to be doing, will prevent correct testing of the case(s) where more than one address group exists.

Reply:
@zeptin ah with this, I'm also a bit confused on what the outcome should look like. I made a few transactions with a test net version of Core:

```
"Date","Type","Label","Address","Amount (BTC)","ID"
"2019-03-28T14:42:36","Sent to","Faucet","2N6DXYQsnVYLRof79KF9L98qb1419uADHN9","-0.09000256","e1c6a34432ac646fd96b14c46becce355f36ba2cfda810d6dc079e4ed1ef21e7"
"2019-03-27T13:19:50","Sent to","Faucet","2N6DXYQsnVYLRof79KF9L98qb1419uADHN9","-0.01000166","d59a409a35bdab8e08a1ffe3123fc38b183e58c837b5ad88504d7e8b4390abbe"
"2019-03-27T13:18:49","Received with","","2MsvSRsRbmWjzDXJ1BKrXXL3eb22JPPYsv5","0.01000000","7703be808c3c9ca936109f28aa17d2624e5da8d8a84344644c596e5d0a9745db"
"2019-03-27T11:44:50","Sent to","Faucet","2N6DXYQsnVYLRof79KF9L98qb1419uADHN9","-0.10000166","3525b1d4fc2e04e740acf79f75516e5dbd09fc304067321aab960bf8f278d814"
"2019-03-27T10:44:49","Received with","Faucet","2MxWJcKtqDvNMuo5e6kgtSkAmiZBfz1ne2v","0.19672475","7b2742e30cc1de6337900c959eea57bfea8a64cd68448687bbbd687fc235b54f"
```

and the output from `listaddressgroupings` was this:

```
[
  [
    [
      "2MsvSRsRbmWjzDXJ1BKrXXL3eb22JPPYsv5",
      0.00000000
    ],
    [
      "2MxCDdJ9RTqLUsirPm6D9uY2mfsEQFN6HJf",
      0.00000000
    ],
    [
      "2MxWJcKtqDvNMuo5e6kgtSkAmiZBfz1ne2v",
      0.00000000,
      "Faucet"
    ],
    [
      "2NC3h3Twgrvm22cwiN5pYcUXqEUCJihfL81",
      0.00671887
    ]
  ]
]
```

As you can see, they also list a list of groupings? I think that the term "grouping" here is confusing or inappropriate because from my finding I am not sure how it is possible to "group" addresses like this?

If you look at the 2nd transaction (the are ordered from latest to oldes) I sent 0.1 BTC to another wallet. The changes address that was generated was `2MsvSRsRbmWjzDXJ1BKrXXL3eb22JPPYsv5`, 

I then sent 0.1 to the change address, hoping this create a "grouping" but it did not. I just added the change address as another entry to the list of results.

Personally, I think the purpose of `listaddressgroupings` is to list addresses (including change addresses) that has been used to send or receive funds in the wallet and give their respective balances at a given point in time.

@mikedennis @zeptin thoughts?

Reply:
OK upon further investigation, the next set of groupings happen when we send funds to a new unused address in the wallet:

```
[
  [
    [
      "2MsvSRsRbmWjzDXJ1BKrXXL3eb22JPPYsv5",
      0.00000000
    ],
    [
      "2MxCDdJ9RTqLUsirPm6D9uY2mfsEQFN6HJf",
      0.00000000
    ],
    [
      "2MxWJcKtqDvNMuo5e6kgtSkAmiZBfz1ne2v",
      0.00000000,
      "Faucet"
    ],
    [
      "2NC3h3Twgrvm22cwiN5pYcUXqEUCJihfL81",
      0.00671887
    ]
  ],
  [
    [
      "2MwCR8ii5cBBPNx8NDehpypJnarphmmUL2p",
      0.05000000,
      "Block.IO"
    ]
  ]
```

I will update the code and tests accordingly.